oap-server/server-core/src/main/java/org/apache/skywalking/oap/server/core/profile/analyze/ProfileStackNode.java

```diff
@@ -0,0 +1,268 @@
+        }
+
+        for (ProfileStackNode childrenNode : targetNode.children) {
+            // find node from being merged node children
+            for (ListIterator<ProfileStackNode> it = beingMergedNode.children.listIterator(); it.hasNext();) {
+                ProfileStackNode node = it.next();
+                if (node != null && node.matches(childrenNode)) {
+                    childrenNode.combineDetectedStacks(node);
+                    continueChildrenMerging.accept(new Pair<>(childrenNode, node));
+
+                    it.set(null);
+                    break;
+                }
+            }
+        }
+
+        // targetNode.children add to beingMergedNode.children, merge children to single list
+        int searchingIndex = 0;
+        int beingMergeChildrenCount = beingMergedNode.children.size();
+        for (int nodeInx = 0; nodeInx < targetNode.children.size(); ) {
+            if (searchingIndex >= beingMergeChildrenCount) {
+                beingMergedNode.children.addAll(targetNode.children.subList(nodeInx, targetNode.children.size()));
+                break;
+            }
+            for (; searchingIndex < beingMergeChildrenCount; searchingIndex++) {
+                if (beingMergedNode.children.get(searchingIndex) == null) {
+                    beingMergedNode.children.set(searchingIndex, targetNode.children.get(nodeInx));
+                    nodeInx++;
+                    break;
+                }
+            }
+        }
+        targetNode.children = beingMergedNode.children;
+    }
+
+    /**
+     * build GraphQL result, calculate duration and count data using parallels
+     * @return
+     */
+    public ProfileStackElement buildAnalyzeResult() {
+        // all nodes add to single-level list (such as flat), work for parallel calculating
+        LinkedList<Pair<ProfileStackElement, ProfileStackNode>> nodeMapping = new LinkedList<>();
+        ProfileStackElement root = buildElement();
+        nodeMapping.add(new Pair<>(root, this));
+
+        // same with combine logic
+        LinkedList<Pair<ProfileStackElement, ProfileStackNode>> stack = new LinkedList<>();
+        stack.add(new Pair<>(root, this));
+        while (!stack.isEmpty()) {
+            Pair<ProfileStackElement, ProfileStackNode> needCombineNode = stack.pop();
+            ProfileStackElement combineTo = needCombineNode.key;
+
+            // generate children node and add to stack and all node mapping
+            combineTo.setChilds(needCombineNode.value.children.stream().map(c -> {
+                ProfileStackElement element = c.buildElement();
+                Pair<ProfileStackElement, ProfileStackNode> pair = new Pair<>(element, c);
+                stack.add(pair);
+                nodeMapping.add(pair);
+
+                return element;
+            }).collect(Collectors.toList()));
+        }
+
+        // calculate durations
+        nodeMapping.parallelStream().forEach(t -> t.value.calculateDuration(t.key));
```

Comment:
Why require `ProfileStackNode`(t.value) to calculate duration? Using List<ProfileStackElement> should be enough